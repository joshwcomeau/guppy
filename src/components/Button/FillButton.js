// @flow
import React, { Component } from 'react';

import { COLORS, GRADIENTS } from '../../constants';

import ButtonBase from './ButtonBase';

type Props = {
  colors?: Array<string>,
  hoverColors?: Array<string>,
  textColor: string,
  children: React$Node,
  disabled?: boolean,
};

export const wrapColorsInGradient = (colors?: Array<string> | string) => {
  if (!Array.isArray(colors)) {
    return colors;
  }

  if (colors.length === 1) {
    return colors[0];
  }

  return `linear-gradient(
    45deg,
    ${colors.join(',')}
  )`;
};

class FillButton extends Component<Props> {
  static defaultProps = {
    colors: GRADIENTS.primary,
    textColor: COLORS.textOnBackground,
  };

  render() {
    const { colors, hoverColors, children, ...delegated } = this.props;

    const background = wrapColorsInGradient(colors);
    const hoverBackground = wrapColorsInGradient(hoverColors) || background;

    return (
      <ButtonBase
        activeSplat
        background={background}
        hoverBackground={hoverBackground}
        {...delegated}
      >
        {children}
      </ButtonBase>
    );
  }
}

export default FillButton;
