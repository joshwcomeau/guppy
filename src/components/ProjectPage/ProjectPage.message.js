// @flow
import { defineMessages } from 'react-intl';

const messages = {
  fail: {
    ohNo: {
      id: 'projectPage.fail.ohNo',
      defaultMessage: 'Oh no!',
      description: 'Text for "Oh no!" when project fails',
    },
    missing: {
      id: 'projectPage.fail.missing',
      defaultMessage: ' Looks like your project dependencies are missing.',
      description:
        'Text to show the project dependencies are missing when project fails',
    },
    freshClone: {
      id: 'projectPage.fail.freshClone',
      defaultMessage:
        "This can happen if you've freshly cloned a project from GitHub." +
        " In order to run scripts in Guppy, you'll need to install them now.",
      description:
        "Text to show the the fail might happen if it's a fresh clone" +
        ' and prompt the user to install the dependencies',
    },
    install: {
      id: 'projectPage.fail.install',
      defaultMessage: 'Install Dependencies',
      description: 'Label for the "Install Dependencies" button',
    },
  },
  main: {
    openInEditor: {
      id: 'projectPage.main.openInEditor',
      defaultMessage: 'Open in Editor',
      description: 'Label for the "Open in Editor" button',
    },
  },
};

const definedMessages = defineMessages(messages);

export default definedMessages;
