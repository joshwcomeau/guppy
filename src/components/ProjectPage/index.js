// @flow
export { default } from './ProjectPage';
