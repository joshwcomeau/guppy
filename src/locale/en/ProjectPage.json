{
	"projectPage.fail.ohNo": "Oh no!",
	"projectPage.fail.freshClone": "This can happen if you've freshly cloned a project from GitHub. In order to run scripts in Guppy, you'll need to install them now.",
	"projectPage.fail.install": "Install Dependencies",
	"projectPage.fail.missing": " Looks like your project dependencies are missing.",
	"projectPage.main.openInEditor": "Open in Editor"
}
