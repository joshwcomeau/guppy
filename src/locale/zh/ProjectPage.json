{
	"projectPage.fail.ohNo": "哦不！",
	"projectPage.fail.freshClone": "这有可能是因为您刚刚从Github上clone了项目。为了在Guppy中运行脚本，现在您需要安装它们",
	"projectPage.fail.install": "安装依赖",
	"projectPage.fail.missing": " 看起来您缺少一些项目依赖",
	"projectPage.main.openInEditor": "在编辑器中打开"
}
