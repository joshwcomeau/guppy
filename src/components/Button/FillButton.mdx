---
name: FillButton
menu: Reusable Components
---

import { Playground, Props } from 'docz';

import { RAW_COLORS } from '../../constants';
import FillButton from './FillButton';

# FillButton

<Props of={FillButton} />

## Basic usage

<Playground>
  <FillButton>Default</FillButton>
</Playground>

## Sizes

<Playground>
  <FillButton size="xsmall">
    Extra-small button
  </FillButton>
</Playground>

<Playground>
  <FillButton size="small">
    Small button
  </FillButton>
</Playground>

<Playground>
  <FillButton size="medium">
    Medium button
  </FillButton>
</Playground>

<Playground>
  <FillButton size="large">
    Large button
  </FillButton>
</Playground>

## Colors

<Playground>
  <FillButton colors={[RAW_COLORS.green[500], RAW_COLORS.green[500]]}>
    Solid green
  </FillButton>
</Playground>

<Playground>
  <FillButton colors={[RAW_COLORS.orange[500], RAW_COLORS.pink[300]]}>
    Orange to pink
  </FillButton>
</Playground>

<Playground>
  <FillButton colors={[RAW_COLORS.teal[700], RAW_COLORS.lime[500]]}>
    Teal to lime
  </FillButton>
</Playground>

## Disabled

<Playground>
  <FillButton disabled>
  FillButton
  </FillButton>
</Playground>
