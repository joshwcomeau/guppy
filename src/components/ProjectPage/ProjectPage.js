// @flow
import React, { PureComponent, Fragment } from 'react';
import { connect } from 'react-redux';
import styled, { keyframes } from 'styled-components';
import { Tooltip } from 'react-tippy';

import { getSelectedProject } from '../../reducers/projects.reducer';
import { COLORS } from '../../constants';
import * as actions from '../../actions';

import MainContentWrapper from '../MainContentWrapper';
import Heading from '../Heading';
import PixelShifter from '../PixelShifter';
import Spacer from '../Spacer';
import { FillButton } from '../Button';
import DevelopmentServerPane from '../DevelopmentServerPane';
import TaskRunnerPane from '../TaskRunnerPane';
import DependencyManagementPane from '../DependencyManagementPane';
import SettingsButton from '../SettingsButton';
import {
  openProjectInEditor,
  openProjectInFolder,
} from '../../services/shell.service';
import { getCopyForOpeningFolder } from '../../services/platform.service';

import type { Project } from '../../types';
import type { Dispatch } from '../../actions/types';

type Props = {
  project: Project,
  loadDependencyInfoFromDisk: Dispatch<
    typeof actions.loadDependencyInfoFromDiskStart
  >,
};

class ProjectPage extends PureComponent<Props> {
  openIDE = () => {
    const { project } = this.props;
    openProjectInEditor(project);
  };

  openFolder = () => {
    const { project } = this.props;
    // Show a folder in the file manager
    openProjectInFolder(project);
  };

  componentDidMount() {
    const { project, loadDependencyInfoFromDisk } = this.props;

    window.scroll({
      top: 0,
      left: 0,
      behavior: 'smooth',
    });

    loadDependencyInfoFromDisk(project.id, project.path);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.project.id !== nextProps.project.id) {
      this.props.loadDependencyInfoFromDisk(
        nextProps.project.id,
        nextProps.project.path
      );
    }
  }

  render() {
    const { project } = this.props;
    const { path } = project;

    return (
      <FadeIn>
        <MainContentWrapper>
          <FlexRow>
            <PixelShifter
              x={-2}
              reason="Align left edge of title with the modules on page"
            >
              <Tooltip title={path} position="bottom">
                <Heading size="xlarge" style={{ color: COLORS.purple[500] }}>
                  {project.name}
                </Heading>
              </Tooltip>
            </PixelShifter>
            <SettingsButton />
          </FlexRow>

          <ProjectActionBar>
            <FillButton
              colors={COLORS.gray[200]}
              hoverColors={COLORS.gray[300]}
              textColor={COLORS.gray[900]}
              size="small"
              onClick={this.openFolder}
            >
              {getCopyForOpeningFolder()}
            </FillButton>
            <Spacer size={15} />
            <FillButton
              colors={COLORS.gray[200]}
              hoverColors={COLORS.gray[300]}
              textColor={COLORS.gray[900]}
              size="small"
              onClick={this.openIDE}
            >
              Open in Editor
            </FillButton>
          </ProjectActionBar>

          <Spacer size={30} />
          <DevelopmentServerPane leftSideWidth={300} />

          <Spacer size={30} />
          <TaskRunnerPane leftSideWidth={200} />

          {project.dependencies.length > 0 && (
            <Fragment>
              <Spacer size={30} />
              <DependencyManagementPane />
            </Fragment>
          )}

          <Spacer size={60} />
        </MainContentWrapper>
      </FadeIn>
    );
  }
}

const FlexRow = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const ProjectActionBar = styled.div`
  display: flex;
`;

const fadeIn = keyframes`
  from { opacity: 0.5 }
  to { opacity: 1 }
`;

const FadeIn = styled.div`
  animation: ${fadeIn} 400ms;
`;

const mapStateToProps = state => ({
  project: getSelectedProject(state),
});

export default connect(
  mapStateToProps,
  {
    loadDependencyInfoFromDisk: actions.loadDependencyInfoFromDiskStart,
  }
)(ProjectPage);
