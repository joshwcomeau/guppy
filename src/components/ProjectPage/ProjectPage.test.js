/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { shallow } from 'enzyme';

import { ProjectPage, FlexRow } from './ProjectPage';
import FillButton from '../Button/FillButton';

import {
  openProjectInEditor,
  openProjectInFolder,
} from '../../services/shell.service';

jest.mock('../../services/shell.service');

jest.mock('../../services/platform.service', () => ({
  getCopyForOpeningFolder: () => 'Open folder',
}));

const initialState = {
  projects: {
    byId: {
      'a-project': {
        path: 'path/to/project',
        guppy: {
          id: 'a-project',
          name: 'example',
          type: 'create-react-app',
          color: '#fff',
        },
      },
    },
    selectedId: 'a-project',
  },
};

const project = {
  id: 'a-project', // Todo: Check how the id is added in app to project prop
  ...initialState.projects.byId['a-project'],
};

window.scroll = jest.fn();

describe('ProjectPage component', () => {
  let wrapper;
  let instance;
  let mockActions;

  const loadingStates = ['loading', 'done', 'fail'];
  const mountWithStatus = (status, actions) =>
    shallow(
      <ProjectPage
        project={project}
        dependenciesLoadingStatus={status}
        {...actions}
      />
    );

  beforeEach(() => {
    mockActions = {
      loadDependencyInfoFromDisk: jest.fn(),
      reinstallDependencies: jest.fn(),
    };
    wrapper = mountWithStatus('loading', mockActions);
    instance = wrapper.instance();
  });

  describe('Rendering', () => {
    loadingStates.forEach(dependencyLoadingStatus => {
      it(`should render project page (${dependencyLoadingStatus})`, () => {
        wrapper = mountWithStatus(dependencyLoadingStatus, mockActions);
        expect(wrapper.instance().renderConditionally()).toMatchSnapshot();
      });
    });
  });

  it('should render header FlexRow', () => {
    expect(wrapper.find(FlexRow)).toMatchSnapshot();
  });

  it('should dispatch loadDependencyInfoFromDisk on mount', () => {
    expect(mockActions.loadDependencyInfoFromDisk).toHaveBeenCalledWith(
      'a-project',
      project.path
    );
  });

  it('should scroll window to top on mount', () => {
    // Note: Would be nice if we would export the object literal so we can use it here
    expect(window.scroll).toHaveBeenCalledWith({
      top: 0,
      left: 0,
      behavior: 'smooth',
    });
  });

  it('should load dependencies for new project', () => {
    instance.componentWillReceiveProps({
      project: {
        ...project,
        id: 'new-project',
      },
    });
    expect(mockActions.loadDependencyInfoFromDisk).toHaveBeenCalledTimes(2);
  });

  it('should not load dependency if no new project', () => {
    instance.componentWillReceiveProps({
      project,
      dependenciesLoadingStatus: 'done',
    });
    expect(mockActions.loadDependencyInfoFromDisk).toHaveBeenCalledTimes(1);
  });

  it('should trigger open editor', () => {
    instance.openIDE();
    expect(openProjectInEditor).toHaveBeenCalledWith(project);
  });

  it('should trigger open folder', () => {
    instance.openFolder();
    expect(openProjectInFolder).toHaveBeenCalledWith(project);
  });

  it('should trigger reinstall on click (if node_modules missing)', () => {
    const reinstallDependenciesMock = jest.fn();
    wrapper.setProps({
      dependenciesLoadingStatus: 'fail',
      reinstallDependencies: reinstallDependenciesMock,
    });
    const button = wrapper.find(FillButton).last();
    button.simulate('click');

    expect(reinstallDependenciesMock).toHaveBeenCalled();
  });
});
