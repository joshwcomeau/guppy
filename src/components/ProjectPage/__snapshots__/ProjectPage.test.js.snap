// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPage component Rendering should render project page (done) 1`] = `
<React.Fragment>
  <Connect(DevelopmentServerPane)
    leftSideWidth={300}
  />
  <ForwardRef
    size={30}
  />
  <Connect(TaskRunnerPane)
    leftSideWidth={200}
  />
  <ForwardRef
    size={30}
  />
  <Connect(DependencyManagementPane) />
</React.Fragment>
`;

exports[`ProjectPage component Rendering should render project page (fail) 1`] = `
<ForwardRef>
  <ForwardRef>
    <ForwardRef />
  </ForwardRef>
  <ForwardRef>
    <strong>
      Oh no!
    </strong>
     Looks like your project dependencies are missing.
  </ForwardRef>
  <ForwardRef>
    This can happen if you've freshly cloned a project from GitHub. In order to run scripts in Guppy, you'll need to install them now.
  </ForwardRef>
  <ForwardRef
    size={30}
  />
  <ForwardRef>
    <FillButton
      colors={
        Array [
          "#00C853",
          "#69db0d",
        ]
      }
      onClick={[Function]}
      size="large"
      textColor="#FFF"
    >
      Install Dependencies
    </FillButton>
  </ForwardRef>
</ForwardRef>
`;

exports[`ProjectPage component Rendering should render project page (loading) 1`] = `
<MountAfter
  delay={100}
  reason="Don't show spinner if we're really fast."
>
  <ForwardRef>
    <Spinner
      size={50}
    />
  </ForwardRef>
</MountAfter>
`;

exports[`ProjectPage component should render header FlexRow 1`] = `
<styled.div>
  <PixelShifter
    reason="Align left edge of title with the modules on page"
    x={-2}
  >
    <Tooltip
      animateFill={true}
      animation="shift"
      arrow={false}
      arrowSize="regular"
      className=""
      delay={0}
      disabled={false}
      distance={10}
      duration={375}
      followCursor={false}
      hideDelay={0}
      hideDuration={375}
      hideOnClick={true}
      html={null}
      inertia={false}
      interactive={false}
      interactiveBorder={2}
      multiple={false}
      offset={0}
      onHidden={[Function]}
      onHide={[Function]}
      onRequestClose={[Function]}
      onShow={[Function]}
      onShown={[Function]}
      popperOptions={Object {}}
      position="bottom"
      size="regular"
      sticky={false}
      stickyDuration={200}
      style={Object {}}
      theme="dark"
      title="path/to/project"
      touchHold={false}
      trigger="mouseenter focus"
      unmountHTMLWhenHide={false}
    >
      <Heading
        size="xlarge"
        style={
          Object {
            "color": "#651fff",
          }
        }
      />
    </Tooltip>
  </PixelShifter>
  <Connect(SettingsButton) />
</styled.div>
`;
